File: backend/app/api/reasoning_flow.py
"""API endpoints for reasoning flow (推理流可视化配置).

P0：直接改 active 流的 node_config（立即生效）；版本切版 API 预留给二期 draft 试错流程。
三层兜底在 run_pipeline（DB 异常回退模块常量），API 层不兜底。
"""
import logging
from fastapi import APIRouter, HTTPException
from app.repository.reasoning_flow_repo import ReasoningFlowRepository

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/api/reasoning-flow", tags=["reasoning-flow"])

_VALID_NODE_KEYS = {"extract", "select_baseline", "match_kp", "compose", "review", "condition", "ask_user", "clarity_check", "budget_check", "scene_analysis", "cond_scene", "normalize_input", "confirm_series", "llm_understand", "slot_validate", "confirm", "llm_ask", "llm_audit"}


def _is_valid_node_key(key: str) -> bool:
    """节点 key 校验：固定 key（extract/scene_analysis/…）或画布 palette 新增节点的后缀 id
    （addNode 生成 extract_1 / scene_analysis_2，executor 按节点 id 读 config）。
    只认合法 base 类型，防止任意 key 写入。"""
    if key in _VALID_NODE_KEYS:
        return True
    base = key.rsplit("_", 1)[0]
    return base in _VALID_NODE_KEYS


@router.get("/")
def get_active():
    """取 active 流（含 graph + node_configs 按 node_key 索引）。无 active 返回 {flow: null}。"""
    repo = ReasoningFlowRepository()
    try:
        return {"flow": repo.get_active_flow()}
    finally:
        repo.close()


# 带 LLM 开关的节点类型（enable_llm 开关在这些节点上生效）
LLM_NODE_TYPES = {"llm_understand", "llm_audit", "llm"}


@router.get("/llm-nodes")
def list_llm_nodes():
    """列出 active 流中所有带 LLM 开关的节点（llm_understand/llm_audit/llm）及当前开关状态。

    供需求分析页「LLM 节点」按钮：一次看清哪些节点挂了 LLM、各自开关状态。
    """
    repo = ReasoningFlowRepository()
    try:
        f = repo.get_active_flow()
        if not f:
            return {"nodes": []}
        cfg_map = f.get("node_configs") or {}
        out = []
        for n in (f.get("graph") or {}).get("nodes") or []:
            ntype = n.get("type") or n.get("id")
            if ntype in LLM_NODE_TYPES:
                cfg = cfg_map.get(n.get("id")) or {}
                out.append({
                    "id": n.get("id"),
                    "type": ntype,
                    "label": n.get("label") or n.get("id"),
                    "enable_llm": bool(cfg.get("enable_llm")),
                })
        return {"nodes": out}
    finally:
        repo.close()


@router.get("/versions")
def list_versions():
    repo = ReasoningFlowRepository()
    try:
        return {"versions": repo.list_versions()}
    finally:
        repo.close()


@router.put("/graph")
def update_graph(data: dict):
    """改 active 流图结构（一期改坐标/标签；二期拖拽编排接 stencil+dnd）。"""
    graph = data.get("graph")
    if not isinstance(graph, dict):
        raise HTTPException(400, "Missing graph")
    repo = ReasoningFlowRepository()
    try:
        f = repo.get_active_flow()
        if not f:
            raise HTTPException(404, "No active reasoning flow")
        return repo.upsert_graph(f["id"], graph, operator=data.get("operator", "system"))
    finally:
        repo.close()


@router.put("/nodes/{node_key}")
def update_node(node_key: str, data: dict):
    """改 active 流某节点 config（立即生效：下次推理即用新参数）。"""
    if not _is_valid_node_key(node_key):
        raise HTTPException(400, f"Invalid node_key: {node_key}")
    config = data.get("config")
    if not isinstance(config, dict):
        raise HTTPException(400, "Missing config")
    repo = ReasoningFlowRepository()
    try:
        f = repo.get_active_flow()
        if not f:
            raise HTTPException(404, "No active reasoning flow")
        return repo.upsert_node_config(f["id"], node_key, config, operator=data.get("operator", "system"))
    finally:
        repo.close()


@router.post("/versions/{flow_id}/activate")
def activate(flow_id: int, data: dict = None):
    repo = ReasoningFlowRepository()
    try:
        f = repo.activate(flow_id, operator=(data or {}).get("operator", "system"))
        if not f:
            raise HTTPException(404, f"Flow {flow_id} not found")
        return f
    finally:
        repo.close()


@router.post("/test-run")
async def test_run(body: dict):
    """试运行 playground：输入需求文本（+可选预算），同步跑 active flow 图执行器，
    返回每步事件 + ext/kp_by_model/plans 明细。供策略中心画布编辑器交互测试。

    - 不绑商机（opportunity_id 传占位 "test-run"——run_graph_executor 内部从不引用它）。
    - force_complete 默认 True（跳过反问、一键出方案）；前端可传 False 测反问补全（unclear/partial 会暂停在 ask_user）。
    - 不回退 linear fallback：调试工具，图执行的报错原样暴露给用户看（仅包一层 except 返回 error+events）。
    - 明细全从 ctx 取（step_done 的 payload 是摘要级，明细在 ctx.kp_by_model / ctx.plans）。
    """
    text = (body or {}).get("requirement_text")
    if not text:
        raise HTTPException(400, "Missing requirement_text")
    budget = (body or {}).get("explicit_budget")
    force_complete = bool((body or {}).get("force_complete", True))
    repo = ReasoningFlowRepository()
    try:
        flow = repo.get_active_flow()
    finally:
        repo.close()
    if not flow:
        raise HTTPException(400, "无 active 推理流，请先在画布配置节点")

    events: list = []

    async def _collect(payload: dict):
        events.append(payload)

    from app.services.reasoning_executor import run_graph_executor
    initial_ctx = {"budget": budget, "force_complete": force_complete}
    try:
        ctx = await run_graph_executor(
            "test-run", text, flow, _collect, initial_ctx=initial_ctx
        )
    except Exception as e:
        logger.exception("test-run 图执行失败")
        return {"error": str(e), "events": events}

    return {
        "events": events,
        "ext": ctx.get("ext") or {},
        "kp_by_model": ctx.get("kp_by_model") or {},
        "plans": ctx.get("plans") or [],
        "awaiting_input": bool(ctx.get("awaiting_input")),
    }

File: frontend/src/api/reasoningFlow.ts
/**
 * 推理流可视化配置 API（/api/reasoning-flow）。
 * 推理流 DAG：5 节点 extract→select_baseline→match_kp→compose→review + 各节点参数 config。
 * 改节点 config 立即生效（下次推理用新参数）；三层兜底在 run_pipeline（DB 异常回退模块常量）。
 */
import axios from 'axios'
import type { Plan } from '@/api/reasoning'

const RESP = <T>(p: Promise<{ data: T }>) => p.then(r => r.data)

export type ReasoningNodeKey = 'extract' | 'select_baseline' | 'match_kp' | 'compose' | 'review' | 'ask_user' | 'clarity_check' | 'budget_check' | 'scene_analysis' | 'cond_scene' | 'normalize_input' | 'llm_understand' | 'slot_validate' | 'llm_audit'

/** extract 节点配置（多词表体系：KP 表 + 机型表，左侧 DB 下拉动态） */
export interface LexiconEntry {
  key: string                                    // 左侧下拉选中的值（品类名 / 系列 / 形态 / 类型名）
  triggers: string[]                             // 右侧触发词
}
export interface Lexicon {
  id: string                                     // 词表 id（lex_kp / lex_model）
  name: string                                   // 词表显示名
  kind: 'kp' | 'model'                           // 决定左侧下拉数据源
  entries: LexiconEntry[]
}
export interface ExtractNodeConfig {
  keyword_limit?: number
  lexicons?: Lexicon[]                           // 新：多词表（两张主表）
  // 旧字段保留兼容已存配置；编辑器检测到旧结构自动转新
  category_lexicon?: Record<string, string[]>
  series_keywords?: string[]
  series_keyword_map?: Record<string, string>
  stopwords?: string[]
  engine_note?: string
}

/** review 节点配置（P6 产出形态；BOM 模板不在此节点） */
export interface ReviewNodeConfig {
  output_preset?: 'detailed' | 'standard' | 'concise'
  output_fields?: {
    show_price?: boolean
    merge_chassis_kp?: boolean
    currency?: string
    show_recommend_reason?: boolean
    show_missing_hint?: boolean
  }
}

export interface ReasoningNodeMeta {
  id: string
  type: string
  label: string
  position?: { x: number; y: number }
}
export interface ReasoningGraph {
  nodes: ReasoningNodeMeta[]
  edges: Array<{
    id?: string
    source: string
    target: string
    source_handle?: string | null
    target_handle?: string | null
  }>
}

export interface ReasoningFlow {
  id: number
  name: string
  version: number
  status: string
  graph: ReasoningGraph
  is_active: boolean
  description: string | null
  node_configs?: Partial<Record<ReasoningNodeKey, Record<string, any>>>  // 仅 get_active 返回
}

/** 试运行事件（按执行顺序：pipeline_start / step_start / step_done / candidates_ready / pipeline_done / need_input） */
export interface TestRunEvent {
  type: string
  step?: string
  label?: string
  payload?: any
  [k: string]: any
}
/** 试运行结果：每步事件 + ext/kp_by_model/plans 明细（全从 ctx 取） */
export interface TestRunResult {
  events: TestRunEvent[]
  ext: Record<string, any>
  kp_by_model: Record<string, any[]>
  plans: Plan[]
  awaiting_input: boolean
  error?: string
}

export const reasoningFlowApi = {
  get: () => RESP<{ flow: ReasoningFlow | null }>(axios.get('/api/reasoning-flow/')),
  listVersions: () => RESP<{ versions: ReasoningFlow[] }>(axios.get('/api/reasoning-flow/versions')),
  updateGraph: (graph: ReasoningGraph) =>
    RESP<ReasoningFlow>(axios.put('/api/reasoning-flow/graph', { graph })),
  updateNode: (nodeKey: ReasoningNodeKey, config: Record<string, any>) =>
    RESP<any>(axios.put(`/api/reasoning-flow/nodes/${nodeKey}`, { config })),
  activate: (flowId: number) =>
    RESP<ReasoningFlow>(axios.post(`/api/reasoning-flow/versions/${flowId}/activate`, {})),
  llmNodes: () => RESP<{ nodes: Array<{ id: string; type: string; label: string; enable_llm: boolean }> }>(axios.get('/api/reasoning-flow/llm-nodes')),
  testRun: (text: string, budget?: number, forceComplete?: boolean) =>
    RESP<TestRunResult>(axios.post('/api/reasoning-flow/test-run', {
      requirement_text: text,
      explicit_budget: budget,
      force_complete: forceComplete ?? true,
    })),
}
